### PR TITLE
feat(dummy diag publisher): change diag name specification method to YAML

### PR DESCRIPTION
**Pull request by asana17**
Signed-off-by: asana17 [akihiro.sakurai@tier4.jp](mailto:akihiro.sakurai@tier4.jp)
## Description
`diag_name` parameter of the ROS node `dummy_diag_publisher` is configured in the node launch file.  `dummy_diag_publisher` can take only one `diag name` parameter and the ROS node is launched for each diagnostic. This causes differences in the launch files among projects.  
I changed to publish diagnostic messages in one ROS node and multiple `diag_name` parameters can be configured in the parameter YAML file. `status` and `is_active` parameters of each diagnostic can be reconfigured from the command line.

I also changed the launch files to update [tier4/autoware_launch](https://github.com/tier4/autoware_launch). Please see the related links below for more details about the launch files.

(I reopened this PR for https://github.com/autowarefoundation/autoware.universe/pull/2745.)

## Changes
1. Use parameters in the YAML file as the name of diagnostics instead of the parameter in the launch file.
2. Use new parameters to read from the YAML file and enable parameter reconfiguration for each diagnostic.
3. Multiple diagnostics messages are published from one ROS node.

## Related links

[TIER IV INTERNAL LINK TO SLACK](https://star4.slack.com/archives/C4P0NSMB5/p1672881973897619)
https://github.com/tier4/autoware_launch/pull/762
https://github.com/autowarefoundation/autoware_launch/pull/206
https://github.com/tier4/aip_launcher/pull/101
https://github.com/autowarefoundation/sample_sensor_kit_launch/pull/55
### dummy_diag_publisher
1. Run dummy_diag_pubilsher and rqt_runtime_monitor. 
    The config YAML file is `system/dummy_diag_publisher/config/dummy_diag_publisher.param.yaml`
  ```
  ros2 launch dummy_diag_publisher dummy_diag_publisher.launch.xml
  ```

 Screenshot from 2023-01-25 15-24-23
  2. Change the parameter from the command line.
  ```
  ros2 param set /dummy_diag_publisher velodyne_temperature.status "Error"
  ```

Screenshot from 2023-01-25 15-25-01

```
ros2 param set /dummy_diag_publisher obstacle_crash.is_active false
```

Screenshot from 2023-01-25 15-26-05

### launch files
I used [tier4/pilot-auto](https://github.com/tier4/pilot-auto) for the tests. I replaced [universe](https://github.com/autowarefoundation/autoware.universe) , [launcher](https://github.com/tier4/autoware_launch), and [aip_launcher](https://github.com/tier4/aip_launcher). I confirmed the planning simulator works successfully in case 1 below.

1. Launch planning simulation. (x1, xx1)

  terminal 1
  ```
  source ~/pilot-auto/install/setup.zsh
  ros2 launch autoware_launch planning_simulator.launch.xml map_path:=/home/asana/autoware_map/sample-map-planning 
  vehicle_model:=lexus sensor_model:=aip_xx1
  ```

  terminal 2
  ```
  ros2 run rqt_runtime_monitor rqt_runtime_monitor
  ```
  ```
  ros2 node list
  ```
  ```
  ros2 topic list
  ```

2. Launch system component (x1, xx1, x2)

 terminal 1
  ```
  ros2 launch autoware_launch tier4_system_component.launch.xml run_mode:=online launch_system_monitor:=true sensor_model:=aip_x2
  ```
  ```
  ros2 launch autoware_launch tier4_system_component.launch.xml run_mode:=planning_simulation launch_system_monitor:=false sensor_model:=aip_x2
  ```
 terminal 2
  ``` 
  ros2 run rqt_runtime_monitor rqt_runtime_monitor
  ```
  ```
  ros2 node list
  ```
  ```
  ros2 topic list
  ```
  
In each case, I compared the rqt_runtime_monitor results, the node list results, and the topic list results and confirmed no difference other than the points below.

- service_log_checker is launched in tier4_system_launch, but is not launched in the previous system_launch.
- The changes of  dummy_diag_publisher nodes: all dummy diags are published by one dummy_diag_publisher node, and the diag names, Hardware IDs are changed as the screenshots.
pilot-auto-xx1-xml-runtimemonitor
pilot-auto-xx1-yaml-runtimemonitor

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by asana17** [comment]
@kminoda I closed the previous PR because I signed off some commits by mistake. Could you check this PR instead of the previous one (https://github.com/autowarefoundation/autoware.universe/pull/2745)?

File: launch/tier4_system_launch/launch/system.launch.xml [hunk 1]
```diff
@@ -85,4 +94,16 @@
       </include>
     </group>
   </group>
+
+  <!-- Dummy Diag Publisher -->
+  <group if="$(eval &quot;'$(var run_mode)'=='planning_simulation'&quot;)">
+    <include file="$(find-pkg-share dummy_diag_publisher)/launch/dummy_diag_publisher.launch.xml">
+      <arg name="config_file" value="$(var dummy_diag_publisher_param_path)"/>
+      <arg name="extra_config_file_sensor" value="$(var sensor_launch_pkg)/config/dummy_diag_publisher/sensor_kit.param.yaml"/>
```
Comment (kminoda) [inline launch/tier4_system_launch/launch/system.launch.xml:102]: I think we also need to add this parameter file in [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch).

**Review by kminoda — COMMENTED** [review]
Thank you for the PR!

File: launch/tier4_system_launch/launch/system.launch.xml (hunk 1 above)
Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: Adding default value (empty param file) to the file above is another way.  Or we need to add an empty param file to [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch)?

Comment (kminoda) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: Both approach is fine for me! (If I had to choose, I would go for the latter one, since setting the default parameter may be confusing)

Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: OK, I will add an empty param file to [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch).

Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @kminoda How do we deal with the gap between  [autowarefoundation/autoware_launch](https://github.com/autowarefoundation/autoware_launch) when we modify [tier4_system_launch/launch/system.launch.xml](https://github.com/autowarefoundation/autoware.universe/pull/2840/files/bd5dac0ddf2299504782cac2773e7f990e332e12#diff-985682f001a2c5fdfbf452fbeb3424fba1093af16bdfd0f618ca228117906adc)?  [autowarefoundation/autoware_launch](https://github.com/autowarefoundation/autoware_launch) does not have dummy_diag_publisher launch implementation. Do we also need empty param files here?

Comment (kminoda) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @asana17 (Sorry, maybe I couldn't get the point) I will leave the decision up to you, as long as planning simulator works fine with awf/autoware and tier4/pilot-auto. I think it's better to first merge this PR ASAP and mitigate those differences later on.

Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @kminoda Modifying [tier4_system_launch/launch/system.launch.xml](https://github.com/autowarefoundation/autoware.universe/pull/2840/files/bd5dac0ddf2299504782cac2773e7f990e332e12#diff-985682f001a2c5fdfbf452fbeb3424fba1093af16bdfd0f618ca228117906adc) also have effects to awf/autoware, because awf/autoware also use this tier4_system_launch/launch/system.launch.xml from awf/autoware_launch. However, awf/autoware_launch does not have a config file corresponding with [here](https://github.com/asana17/autoware.universe/blob/bd5dac0ddf2299504782cac2773e7f990e332e12/launch/tier4_system_launch/launch/system.launch.xml#L101), and we cannot launch the planning simulator (occurs param file path not found error). To deal with this problem, we also need to put an empty param file in awf/autoware_launch.

Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @kminoda Temporarily placing empty param files for dummy_diag_publisher to [sample_sensor_kit_launch](https://github.com/autowarefoundation/sample_sensor_kit_launch) and awf/autoware_launch is better way than setting the default params to tier4_system_launch/launch/system.launch.xml, so I will deal with it.

Comment (kminoda) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @asana17 I see, make sense, thank you :pray: 
I agree with your proposal!

Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @kminoda I made the PRs.  Could you please review them?
https://github.com/autowarefoundation/autoware_launch/pull/206
https://github.com/autowarefoundation/sample_sensor_kit_launch/pull/55

Comment (kminoda) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @asana17 Thanks!
Could you double-check if planning_simulator works in awf/autoware? `dummy_diag_publisher` fails in my case:
```
[ERROR] [dummy_diag_publisher-20]: process has died [pid 1052617, exit code -6, cmd '/home/minoda/ghq/github.com/autowarefoundation/autoware/install/dummy_diag_publisher/lib/dummy_diag_publisher/dummy_diag_publi
sher --ros-args -r __node:=dummy_diag_publisher -p use_sim_time:=False -p wheel_radius:=0.383 -p wheel_width:=0.235 -p wheel_base:=2.79 -p wheel_tread:=1.64 -p front_overhang:=1.0 -p rear_overhang:=1.1 -p left_o
verhang:=0.128 -p right_overhang:=0.128 -p vehicle_height:=2.5 -p max_steer_angle:=0.7 --params-file /home/minoda/ghq/github.com/autowarefoundation/autoware/install/autoware_launch/share/autoware_launch/config/s
ystem/dummy_diag_publisher/dummy_diag_publisher.param.yaml --params-file /home/minoda/ghq/github.com/autowarefoundation/autoware/install/sample_sensor_kit_launch/share/sample_sensor_kit_launch/config/dummy_diag_
publisher/sensor_kit.param.yaml --params-file /tmp/launch_params_4nsgtg10'].

```

Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @kminoda Thank you for checking. I found out that the empty parameter file causes this error. I wanted to treat no param as an error when I implemented the dummy_diag_publisher code, so I temporarily fixed parameter files to set "dummy_diag_empty" diag.

Comment (kminoda) [reply launch/tier4_system_launch/launch/system.launch.xml:102]: @asana17 Thank you! It seems that awf/autoware works.

**Review by kminoda — APPROVED** [review]
LGTM

**Comment by asana17** [comment]
@ito-san @kenji-miyake Could you review this PR?

File: system/dummy_diag_publisher/README.md [hunk 2]
```diff
@@ -16,19 +16,38 @@ This package outputs a dummy diagnostic data for debugging and developing.
 
 ### Node Parameters
 
-| Name          | Type                           | Default Value | Explanation                             | Reconfigurable |
-| ------------- | ------------------------------ | ------------- | --------------------------------------- | -------------- |
-| `update_rate` | int                            | `10`          | Timer callback period [Hz]              | false          |
-| `diag_name`   | string                         | `diag_name`   | Diag_name set by dummy diag publisher   | false          |
-| `is_active`   | bool                           | `true`        | Force update or not                     | true           |
-| `status`      | DummyDiagPublisherNode::Status | `0 (OK)`      | diag status set by dummy diag publisher | true           |
+The parameter `DIAGNOSTIC_NAME` must be a name that exists in the parameter YAML file. If the parameter `status` is given from a commandline, the parameter `is_active` is automatically set to `true`.
```
Comment (ito-san) [inline system/dummy_diag_publisher/README.md:19]: @asana17 Spell check failed in warning. 
> Warning: Unknown word (commandline)
This is not required, but highly recommended.

**Review by ito-san — APPROVED** [review]
LGTM

File: system/dummy_diag_publisher/CMakeLists.txt [hunk 3]
```diff
@@ -4,15 +4,12 @@ project(dummy_diag_publisher)
...
 autoware_package()
 
-ament_auto_add_library(dummy_diag_publisher SHARED
-  src/dummy_diag_publisher_node/dummy_diag_publisher_node.cpp
-)
-
-rclcpp_components_register_node(dummy_diag_publisher
-  PLUGIN "DummyDiagPublisherNode"
-  EXECUTABLE dummy_diag_publisher_node
+ament_auto_add_executable(${PROJECT_NAME}
```
Comment (kenji-miyake) [inline system/dummy_diag_publisher/CMakeLists.txt:7]: Why did you change it to a normal ROS node?
It's generally recommended to use ComposableNode because users can control the process the node launches in.

I'd appreciate it if you could consider changing back to ComposableNode in another PR.

**Review by kenji-miyake — APPROVED** [review]
Thank you for this improvement! And I'm sorry for my late review.
I only added small comments for improvements. It's okay to consider fixing them in another PR.

**Comment by kenji-miyake** [comment]
By the way, it's not a bug fix, I changed the type to `feat.`

File: system/dummy_diag_publisher/CMakeLists.txt (hunk 3 above)
Comment (asana17) [reply system/dummy_diag_publisher/CMakeLists.txt:7]: OK, I will deal with it in another PR.

File: system/dummy_diag_publisher/package.xml [hunk 4]
```diff
@@ -5,13 +5,15 @@
   <version>0.1.0</version>
   <description>The dummy_diag_publisher ROS2 package</description>
   <maintainer email="kenji.miyake@tier4.jp">Kenji Miyake</maintainer>
+  <maintainer email="akihiro.sakurai@tier4.jp">Akihiro Sakurai</maintainer>
```
[the remark this answers is not in the document]
Comment (asana17) [reply system/dummy_diag_publisher/package.xml:8]: I add Ito-san as the maintainer.

File: launch/tier4_system_launch/launch/system.launch.xml [hunk 5]
```diff
@@ -85,4 +94,16 @@
       </include>
     </group>
   </group>
+
+  <!-- Dummy Diag Publisher -->
+  <group if="$(eval &quot;'$(var run_mode)'=='planning_simulation'&quot;)">
```
[the remark this answers is not in the document]
Comment (asana17) [reply launch/tier4_system_launch/launch/system.launch.xml]: I agree with you. I changed the launch flags as you described above.
